### DISCUSSION TITLE
Show something at app start up?

### DISCUSSION THREAD

**Discussion by clawton** [AvaloniaUI/Avalonia#14333] [Q&A]
Hi All,
In a nutshell, is there any way to show something on the screen while the application loads?
(Looked a lot but can't find anything)

We're running an Avalonia based app on a Raspberry PI 4 and we're very happy with it. But one thing that is been thorn in our side is the black screen as the application is starting.
We're using framebuffer based and when our program starts, we immediately run this bit of code: 

```
return builder
    //.UsePlatformDetect()
    .LogToTrace()
    .UseReactiveUI()
    .StartLinuxDrm(args, "/dev/dri/card1");
```
And then we show the main view.  But before the main view shows up, there is a long (multi second) time while presumably the Avalonia framework is doing whatever it does to get ready to show the first view.  Even the boiler plate sample raspberry pi lite code does the same thing.  (On windows, I think this is the time where the application is showing a pink background.)

Here's a link to a video showing the behavior:
https://youtu.be/AAfxYJPcUp0?si=Zdf6i3MwFwKaIqfq

Is there *any way* to show something on the screen instead of just a blank screen? A bitmap would be great.  Even some text...
I tried looking at SkiaSharp stuff but nothing I tried seemed to work.  Probably because something under the covers has control of the screen right away?

We're using Avalonia 0.10.18.

**Comment by thevortexcloud** [comment]
Not sure how well this will work, but you could just write to the frame buffer yourself. It's just a file after all. There is an article [here](https://seenaburns.com/2018/04/04/writing-to-the-framebuffer/) explaining how you can do that as well as how to draw a raw bitmap.

Avalonia may blank the screen though which may clear that all though.

**Reply by clawton** [reply]
Thanks.  Will look into this and if something comes of it, I'll post here. Or close the discussion...
In the meantime, if anyone else has anything to add, please do! :)

**Reply by thevortexcloud** [reply]
One other thing that may work is also just writing to the console. It wont give you anything fancy but you can at least show some sort of initialisation message.

**Reply by clawton** [reply]
Just tried writing random stuff directly to /dev/fb0.

Interesting that if I do it right at start up of the application, I see the random static on the screen.  But then Avalonia grabs the screen and clears it and it stays blank until my first view is shown.

I then put the random writes to fb0 in a forever loop (which included closing and reopening fb0) but no joy.  I saw my "static" at start up but Avalonia gets the screen and takes control.  Nothing in the loop ever fails, just no output.

Another interesting thing is that since it was such a tight loop, I saw the raspberry pi temperature heat warning "Thermometer" icon show up... so something can write to the screen while an Avalonia app is running... :thinking:

Probably something to do with DRM, video cards and such.  I think I'm out of my league... :laughing:

**Comment by timunie** [comment]
1. Use AOT or at least ReadyToRun build on low power devices
2. When loading a lot of stuff during start up, consider to navigate to a "Loading" page first and swtich to MainView after loading completed. Check Avalonia.Samples > Navigation sample for ideas how navigation could work.

**Reply by timunie** [reply]
3. Try upgrading to 11.0 as it has many performance improvements.

**Reply by clawton** [reply]
Thanks.  We're close to release and can't do the upgrade right now.  We're still on dotnet core 6.0 but have plans to do upgrades of dotnet and Avalonia for version 2.0 of the product.  (It's a medical test device so the upgrade is a fairly big deal with tons of regression testing that we will need to redo and can't right now).

**Reply by hisleiterj** [reply]
@clawton I found this discussion because I had the same issue (black screen for around ~6-8 seconds before UI showed up on screen). My setup is a arm64 development board (NXP iMX8ULP, should be in the similar ballpark to a Raspbi 4 in terms of performance) with a small demo app using Avalonia 11.0 and targeting .NET7 using the _ReadyToRun_ option in dotnet publish. So do not expect either the upgrade to Avalonia 11.0 or the usage of the _ReadyToRun_ option to improve start-up time to a degree that the wait time becomes close to something where a splash screen would not be required (your mileage may of course vary). 

However, @timunie suggestion (thanks!) to use AOT (i didn't know that _AOT_ and _ReadyToRun_ are quite different, but they are!) did the trick for me. With _AOT_ and targeting .NET8, my demo app now only takes ~1s to start (which at least for me is fast enough such that no additional splash screen is necessary). Please note: If you are building on a Windows machine it is not supported to use _AOT_ when publishing for Linux arm64 out of the box. While there is no official support from Microsoft for that (c.f. [cross compile](https://learn.microsoft.com/en-us/dotnet/core/deploying/native-aot/cross-compile)), there is a project to work around that, c.f. [PublishAotCross](https://github.com/MichalStrehovsky/PublishAotCross)

**Reply by thevortexcloud** [reply]
The reason why you can't cross compile AOT code is because it needs to be built against the system libraries (after all, it's converting your app into native code instead of CIL). Windows does not have Linux system libraries to build against. Microsoft could work around that with WSL on Windows but they don't seem interested in doing that. In any case, there is nothing stopping you from using WSL yourself to do it.

**Reply by clawton** [reply]
@hisleiterj and @thevortexcloud Thanks for sharing more information and your experiences.  I'm hoping I'll be able to convince our customer to work this into the budget! 😁

**Reply by thevortexcloud** [reply]
Also I should say using AOT is not always possible. It breaks _a lot_ of things. If you do decide to use it, make sure you test it thoroughly.
